File: backend/mob-api/app/Models/Happening.php
<?php

namespace App\Models;

use App\Enums\ActivityLevel;
use App\Enums\HappeningCategory;
use App\Enums\HappeningStatus;
use App\Enums\HappeningType;
use App\Traits\HasUuid;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\HasOne;
use Illuminate\Database\Eloquent\SoftDeletes;

class Happening extends Model
{
    use HasUuid, SoftDeletes;

    protected $fillable = [
        'user_id',
        'title',
        'description',
        'category',
        'type',
        'latitude',
        'longitude',
        'radius_meters',
        'address',
        'starts_at',
        'ends_at',
        'is_ticketed',
        'ticket_price',
        'ticket_quantity',
        'tickets_sold',
        'vibe_score',
        'activity_level',
        'status',
        'hidden_reason',
        'hidden_by',
        'expires_at',
    ];

    protected function casts(): array
    {
        return [
            'category' => HappeningCategory::class,
            'type' => HappeningType::class,
            'status' => HappeningStatus::class,
            'activity_level' => ActivityLevel::class,
            'latitude' => 'float',
            'longitude' => 'float',
            'ticket_price' => 'float',
            'vibe_score' => 'float',
            'is_ticketed' => 'boolean',
            'starts_at' => 'datetime',
            'ends_at' => 'datetime',
            'expires_at' => 'datetime',
        ];
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function snaps(): HasMany
    {
        return $this->hasMany(Snap::class);
    }

    public function tickets(): HasMany
    {
        return $this->hasMany(Ticket::class);
    }

    public function escrow(): HasOne
    {
        return $this->hasOne(Escrow::class);
    }

    public function reports(): HasMany
    {
        return $this->hasMany(Report::class);
    }

    public function hiddenByUser(): BelongsTo
    {
        return $this->belongsTo(User::class, 'hidden_by');
    }

    public function isHidden(): bool
    {
        return in_array($this->status, [HappeningStatus::HIDDEN, HappeningStatus::REPORTED], true);
    }








    public function isHappeningNow(): bool
    {
        if ($this->status !== HappeningStatus::ACTIVE) {
            return false;
        }

        if ($this->expires_at !== null && $this->expires_at->isPast()) {
            return false;
        }


        return $this->starts_at === null || $this->starts_at->isPast();
    }




    public function isUpcoming(): bool
    {
        if ($this->status !== HappeningStatus::ACTIVE) {
            return false;
        }

        return $this->starts_at !== null && $this->starts_at->isFuture();
    }




    public function getDisplayStatus(): string
    {
        if (in_array($this->status, [HappeningStatus::HIDDEN, HappeningStatus::REPORTED], true)) {
            return 'hidden';
        }


        if ($this->status === HappeningStatus::COMPLETED) {
            return 'ended';
        }

        if ($this->status === HappeningStatus::EXPIRED
            || ($this->expires_at !== null && $this->expires_at->isPast())) {
            return 'expired';
        }

        if ($this->starts_at !== null && $this->starts_at->isFuture()) {
            return 'upcoming';
        }

        if ($this->status === HappeningStatus::ACTIVE) {
            return 'live';
        }

        return 'ended';
    }
}

File: mob_app/lib/features/feed/domain/entities/happening.dart
import 'package:equatable/equatable.dart';
import 'package:intl/intl.dart';

import '../../../../shared/models/enums.dart';


class Happening extends Equatable {
  final String uuid;
  final String title;
  final String? description;
  final HappeningCategory category;
  final HappeningType type;
  final double latitude;
  final double longitude;
  final double? radiusMeters;
  final String? address;
  final DateTime? startsAt;
  final DateTime? endsAt;
  final bool isTicketed;
  final double? ticketPrice;
  final int? ticketQuantity;
  final int ticketsSold;
  final int? ticketsRemaining;
  final double vibeScore;
  final ActivityLevel activityLevel;
  final HappeningStatus status;
  final DateTime expiresAt;
  final String? timeRemaining;
  final double? distanceKm;
  final int snapsCount;
  final bool hasSnaps;
  final String? coverImageUrl;


  final String? hostUuid;
  final String? hostName;
  final String? hostAvatarUrl;
  final bool hostIsVerified;
  final String? hostType;

  final DateTime? createdAt;


  final String? displayStatus;


  final bool isActive;


  final bool canBuyTickets;


  final bool canAddSnaps;

  const Happening({
    required this.uuid,
    required this.title,
    this.description,
    required this.category,
    required this.type,
    required this.latitude,
    required this.longitude,
    this.radiusMeters,
    this.address,
    this.startsAt,
    this.endsAt,
    required this.isTicketed,
    this.ticketPrice,
    this.ticketQuantity,
    required this.ticketsSold,
    this.ticketsRemaining,
    required this.vibeScore,
    required this.activityLevel,
    required this.status,
    required this.expiresAt,
    this.timeRemaining,
    this.distanceKm,
    required this.snapsCount,
    required this.hasSnaps,
    this.coverImageUrl,
    this.hostUuid,
    this.hostName,
    this.hostAvatarUrl,
    this.hostIsVerified = false,
    this.hostType,
    this.createdAt,
    this.displayStatus,
    this.isActive = true,
    this.canBuyTickets = false,
    this.canAddSnaps = true,
  });


  bool get isExpired =>
      displayStatus != null
          ? displayStatus == 'expired'
          : DateTime.now().isAfter(expiresAt);


  bool get isUpcoming =>
      displayStatus != null
          ? displayStatus == 'upcoming'
          : startsAt != null && startsAt!.isAfter(DateTime.now());


  bool get isLive =>
      displayStatus != null
          ? displayStatus == 'live'
          : !isExpired && !isUpcoming;


  bool get hasTicketsAvailable =>
      isTicketed &&
      (ticketQuantity == null || ticketsSold < ticketQuantity!);

  static final _priceFormat = NumberFormat('#,##0', 'en_US');


  String get formattedPrice {
    if (!isTicketed || ticketPrice == null) return 'Free';
    return '\u20A6${_priceFormat.format(ticketPrice!.toInt())}';
  }


  String? get formattedDistance {
    if (distanceKm == null) return null;
    if (distanceKm! < 1.0) {
      return '${(distanceKm! * 1000).round()} m';
    }
    return '${distanceKm!.toStringAsFixed(1)} km';
  }


  bool get isAreaBased => radiusMeters != null && radiusMeters! > 0;


  bool get isCasual => type == HappeningType.casual;


  bool get isEvent => type == HappeningType.event;

  @override
  List<Object?> get props => [
        uuid,
        title,
        description,
        category,
        type,
        latitude,
        longitude,
        radiusMeters,
        address,
        startsAt,
        endsAt,
        isTicketed,
        ticketPrice,
        ticketQuantity,
        ticketsSold,
        ticketsRemaining,
        vibeScore,
        activityLevel,
        status,
        expiresAt,
        timeRemaining,
        distanceKm,
        snapsCount,
        hasSnaps,
        coverImageUrl,
        hostUuid,
        hostName,
        hostAvatarUrl,
        hostIsVerified,
        hostType,
        createdAt,
        displayStatus,
        isActive,
        canBuyTickets,
        canAddSnaps,
      ];
}
